src/tree.ts

```diff
@@ -71,38 +71,54 @@ export class Tree {
     return node;
   }
 
-  setter(index: Gindex, expand = false): Link {
-    let link = identity;
+  setNode(gindex: Gindex, n: Node, expand = false): void {
     let node = this.rootNode;
-    const iterator = gindexIterator(index);
-    for (const i of iterator) {
-      if (i) {
-        if (node.isLeaf()) {
-          if (!expand) throw new Error(ERR_INVALID_TREE);
-          else {
-            const child = zeroNode(iterator.remainingBitLength() - 1);
-            node = new BranchNode(child, child);
-          }
+
+    // Pre-compute entire bitstring instead of using an iterator (25% faster)
+    if (gindex < 1) {
+      throw new Error("Invalid gindex < 1");
+    }
+    const bitstring = gindex.toString(2);
+
+    // Keep a list of all parent nodes of node at gindex `index`. Then walk the list
+    // backwards to rebind them "recursively" with the new nodes without using functions
+    const parentNodes: Node[] = [this.rootNode];
+
+    // Ignore the first bit, left right directions are at bits [1,..]
+    // Ignore the last bit, no need to push the target node to the parentNodes array
+    for (let i = 1; i < bitstring.length - 1; i++) {
+      if (node.isLeaf()) {
+        if (!expand) {
+          throw new Error(ERR_INVALID_TREE);
+        } else {
+          node = zeroNode(bitstring.length - i);
```

Comment:
@wemeetagain Can you review this line? How could I test this code path properly?

Reply:
This is for when setting a node at a depth thats further than the existing depth.
The tree would be expanded and filled w zero nodes.

This is what the current behavior looked like:
```
// Lets say we want to insert node n into this tree at index 8
            x
       /         \
      x           x

---
// step one is to expand into zero nodes

            x
       /         \
      z2           x
   /      \
  z1       z1
/    \    /   \
z0   z0  z0   z0

---
// Then replace node at index 8

            x
       /         \
      x           x
   /      \
  x       z1
/    \   /   \
n   z0  z0   z0
```

So to test, you can create a short tree and set a node deeper than the bottom.

Reply:
I'm having trouble creating a test even on master. The behavior seems to be the same between branches tho. I've tested this branch against lodestar, syncing various networks. Will figure out how to test this in a future PR.